pkg/agent/runner.go

```diff
@@ -289,22 +289,35 @@ func (r *Runner) State(ctx context.Context) (*RunnerState, error) {
 
 func (r *Runner) Spawn(ctx context.Context, vmInfoUpdated util.CondChannelReceiver) {
 	go func() {
+		// Trigger restart if necessary *after* we've handled any panics.
+		defer r.global.TriggerRestartIfNecessary(ctx, r.vm.NamespacedName(), r.podIP)
 		// Gracefully handle panics:
 		defer func() {
 			if err := recover(); err != nil {
 				r.setStatus(func(stat *podStatus) {
-					stat.panicked = true
-					stat.done = true
-					stat.errored = fmt.Errorf("runner panicked: %v", err)
+					stat.endState = &podStatusEndState{
+						ExitKind: podStatusExitPanicked,
+						Error:    fmt.Errorf("Runner %v panicked: %v", r.vm.NamespacedName(), err),
```

Comment:
theoretically, it might be nice to use `%w` (but this requires making sure that the recover actually produced a panic) [I wrote some panic handling stuff in `fun` that handles this, but it means that you get yourself an error object that you can introspect usefully.

Reply:
Yeah. Definitely could be nicer, going to leave it for now — I think currently there'd be relatively little benefit from changing it to use `%w`

Reply:
The downside is the implementation complexity — and panic handling in particular benefits from being simple (low code-coverage, things go very wrong if it messes up)